system/Debug/Toolbar.php

```diff
@@ -86,55 +82,226 @@ public function __construct(BaseConfig $config)
 	 *
-	 * @param int $segmentCount
-	 * @param int $segmentDuration
+	 * @param  string $data   JSON encoded Toolbar data
+	 * @param  string $format html, json, xml
 	 *
 	 * @return string
 	 */
-	protected function renderTimeline(int $segmentCount, int $segmentDuration): string
+	protected static function format(string $data, string $format = 'html')
 	{
-		$displayTime = $segmentCount*$segmentDuration;
+		$data = json_decode($data, true);
+
+		// History must be loaded on the fly
+		$filenames = glob(WRITEPATH.'debugbar/debugbar_*');
+		$total     = count($filenames);
+		rsort($filenames);
 
-		$rows = $this->collectTimelineData();
+		$files = [];
+
+		$current = Services::request()->getGet('debugbar_time');
+
+		for ($i = 0; $i < $total; $i++)
+		{
+			// Oldest files will be deleted
+			if ($i >= 30)
+			{
+				unlink($filenames[$i]);
+				continue;
+			}
+
+			// Get the contents of this specific history request
+			ob_start();
+			include($filenames[$i]);
+			$contents = ob_get_contents();
+			ob_end_clean();
+
+			$file = json_decode($contents, true);
+
+			// Debugbar files shown in History Collector
+			$files[$i] = [
+				'time'        => (int)$time = substr($filenames[$i], -10),
+				'datetime'    => date('Y-m-d H:i:s', $time),
+				'active'      => (int)($time == $current),
+				'status'      => $file['vars']['response']['statusCode'],
+				'method'      => $file['method'],
+				'url'         => $file['url'],
+				'isAJAX'      => $file['isAJAX'] ? 'Yes' : 'No',
+				'contentType' => $file['vars']['response']['contentType'],
+			];
+		}
+
+		// Set the History here. Class is not necessary
+		$data['collectors'][] = [
+			'title'           => 'History',
+			'titleSafe'       => 'history',
+			'titleDetails'    => '',
+			'display'         => ['files' => $files],
+			'badgeValue'      => $count = count($files),
+			'isEmpty'         => ! (bool)$count,
+			'hasTabContent'   => true,
+			'hasLabel'        => true,
+			'icon'            => '<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAJySURBVEhL3ZU7aJNhGIVTpV6i4qCIgkIHxcXLErS4FBwUFNwiCKGhuTYJGaIgnRoo4qRu6iCiiIuIXXTTIkIpuqoFwaGgonUQlC5KafU5ycmNP0lTdPLA4fu+8573/a4/f6hXpFKpwUwmc9fDfweKbk+n07fgEv33TLSbtt/hvwNFT1PsG/zdTE0Gp+GFfD6/2fbVIxqNrqPIRbjg4t/hY8aztcngfDabHXbKyiiXy2vcrcPH8oDCry2FKDrA+Ar6L01E/ypyXzXaARjDGGcoeNxSDZXE0dHRA5VRE5LJ5CFy5jzJuOX2wHRHRnjbklZ6isQ3tIctBaAd4vlK3jLtkOVWqABBXd47jGHLmjTmSScttQV5J+SjfcUweFQEbsjAas5aqoCLXutJl7vtQsAzpRowYqkBinyCC8Vicb2lOih8zoldd0F8RD7qTFiqAnGrAy8stUAvi/hbqDM+YzkAFrLPdR5ZqoLXsd+Bh5YCIH7JniVdquUWxOPxDfboHhrI5XJ7HHhiqQXox+APe/Qk64+gGYVCYZs8cMpSFQj9JOoFzVqqo7k4HIvFYpscCoAjOmLffUsNUGRaQUwDlmofUa34ecsdgXdcXo4wbakBgiUFafXJV8A4DJ/2UrxUKm3E95H8RbjLcgOJRGILhnmCP+FBy5XvwN2uIPcy1AJvWgqC4xm2aU4Xb3lF4I+Tpyf8hRe5w3J7YLymSeA8Z3nSclv4WLRyFdfOjzrUFX0klJUEtZtntCNc+F69cz/FiDzEPtjzmcUMOr83kDQEX6pAJxJfpL3OX22n01YN7SZCoQnaSdoZ+Jz+PZihH3wt/xlCoT9M6nEtmRSPCQAAAABJRU5ErkJggg==">',
```

Comment:
This should be just the value of the icon itself. Leave the presentation to the front, since it could be used as an img src, or a background in CSS, etc.

Reply:
All Collectors Icons are being saved with the debugbar file.

One approach could be to define them [here](https://github.com/natanfelles/CodeIgniter4/blob/a59c255fd3c270ec3f5091d3588f26da93baf76f/system/Debug/Toolbar/Views/toolbar.tpl.php#L42) to show the icons only in the HTML output. But all icons would have to be set in the template, which would ignore the `icon()` method defined there in classes.

Or are you referring to remove the img tag and leave only the base64? I didn't quite understand.